diff --git a/nixos/modules/profiles/installation-device.nix b/nixos/modules/profiles/installation-device.nix
index 3a29f7ab8c90..dcc2ddce9ef5 100644
--- a/nixos/modules/profiles/installation-device.nix
+++ b/nixos/modules/profiles/installation-device.nix
@@ -77,7 +77,7 @@ with lib;

         Type `sudo systemctl start display-manager' to
         start the graphical user interface.
-      '';
+      '' + "\n";

     # We run sshd by default. Login is only possible after adding a
     # password via "passwd" or by adding a ssh key to ~/.ssh/authorized_keys.
diff --git a/nixos/modules/services/ttys/getty.nix b/nixos/modules/services/ttys/getty.nix
index c6093034cf75..665825692168 100644
--- a/nixos/modules/services/ttys/getty.nix
+++ b/nixos/modules/services/ttys/getty.nix
@@ -138,7 +138,7 @@ in
     services.getty.greetingLine = mkDefault ''<<< Welcome to ${config.system.nixos.distroName} ${config.system.nixos.label} (\m) - \l >>>'';
     services.getty.helpLine = mkIf (
       config.documentation.nixos.enable && config.documentation.doc.enable
-    ) "\nRun 'nixos-help' for the NixOS manual.";
+    ) "\nRun 'nixos-help' for the NixOS manual.\n";

     systemd.additionalUpstreamSystemUnits = [
       "getty.target"
@@ -199,7 +199,6 @@ in

         [1;32m${config.services.getty.greetingLine}[0m
         ${config.services.getty.helpLine}
-
       '';
     };

