diff --git a/lib/default.nix b/lib/default.nix
index 751738a98965..9a90990fc32a 100644
--- a/lib/default.nix
+++ b/lib/default.nix
@@ -501,6 +501,7 @@ let
         mkAfter
         mkAliasDefinitions
         mkAliasAndWrapDefinitions
+        mkApply
         fixMergeModules
         mkRemovedOptionModule
         mkRenamedOptionModule
diff --git a/lib/modules.nix b/lib/modules.nix
index a206ed6ed8e2..d8eeed1e679e 100644
--- a/lib/modules.nix
+++ b/lib/modules.nix
@@ -1184,9 +1184,35 @@ let
       '';
 
   # Merge definitions of a value of a given type.
-  mergeDefinitions = loc: type: defs: rec {
-    defsFinal' =
+  mergeDefinitions = loc: type: defs: let
+
+    isMkApply = def: def.value._type or "" == "apply";
+    hasMkApply = any isMkApply defsFinal;
+
+    defsApplied = merge: if hasMkApply then
       let
+        fallbackDefs =
+          if length defsFinal == 1 then # prepend the definitions of the next-lower priority:
+            let
+              filtered = (processDefs defsFinal'.highestPrio).defsSorted;
+            in
+              if filtered == [ ] then
+                throw "Option `${showOption loc}' has one definition, but it is a `mkApply' that thus has nothing to be applied to."
+              else
+                filtered ++ defsFinal
+          else
+            defsFinal;
+      in foldl' (prevDefs: def: (
+        # on each "apply", merge the previous defs and continue with the applied value
+        if !(isMkApply def) then prevDefs ++ [ def ] # (the `defs ++ [ def ]` is quadratic in `length defs`)
+        else if prevDefs == [ ] then [ ] # ignore mapper if all values are explicitly ordered after it (see sortProperties)
+        else [ (def // { value = def.value.mapper (merge loc prevDefs); }) ]
+      )) [ ] fallbackDefs
+    else
+      defsFinal;
+
+    processDefs = excludePrio:
+      rec {
         # Process mkMerge and mkIf properties.
         defsNormalized = concatMap (
           m:
@@ -1203,16 +1229,17 @@ let
         ) defs;
 
         # Process mkOverride properties.
-        defsFiltered = filterOverrides' defsNormalized;
+        defsFiltered = filterOverrides'' excludePrio defsNormalized;
 
         # Sort mkOrder properties.
         defsSorted =
           # Avoid sorting if we don't have to.
-          if any (def: def.value._type or "" == "order") defsFiltered.values then
+          if any (def: let type = def.value._type or ""; in type == "order" || type == "apply") defsFiltered.values then
             sortProperties defsFiltered.values
           else
             defsFiltered.values;
-      in
+      };
+    defsFinal' =
       # Fast path: the overwhelming majority of options have exactly one
       # definition whose value carries no property wrapper
       # (mkIf/mkMerge/mkOverride/mkOrder/definition). In that case the
@@ -1236,12 +1263,16 @@ let
           highestPrio = defaultOverridePriority;
         }
       else
+        let inherit (processDefs null) defsSorted defsFiltered; in
         {
           values = defsSorted;
           inherit (defsFiltered) highestPrio;
         };
     defsFinal = defsFinal'.values;
 
+  in rec {
+    inherit defsFinal defsFinal';
+
     # Type-check the remaining definitions, and merge them. Or throw if no definitions.
     mergedValue =
       if isDefined then
@@ -1252,8 +1283,16 @@ let
             throw "A definition for option `${showOption loc}' is not of type `${type.description}'. TypeError: ${checkedAndMerged.headError.message}"
           else
             checkedAndMerged.value
-        else if all (def: type.check def.value) defsFinal then
-          type.merge loc defsFinal
+        else
+
+        if (
+          if hasMkApply then # (this is really just a small performance optimization)
+            all (def: isMkApply def || type.check def.value) defsFinal
+          else
+            all (def: type.check def.value) defsFinal
+        ) then
+            type.merge loc (defsApplied type.merge) # (defsApplied ... == defsFinal without mkApply)
+
         else
           let
             allInvalid = filter (def: !type.check def.value) defsFinal;
@@ -1283,11 +1322,14 @@ let
         (
           if type.merge ? v2 then
             let
+              doMerge = loc: defs: (type.merge.v2 {
+                inherit loc defs;
+              }).value;
               # Check for v2 merge coherence
               r = checkV2MergeCoherence loc type (
                 type.merge.v2 {
                   inherit loc;
-                  defs = defsFinal;
+                  defs = defsApplied doMerge; # (= defsFinal without mkApply)
                 }
               );
             in
@@ -1414,7 +1456,8 @@ let
   */
   filterOverrides = defs: (filterOverrides' defs).values;
 
-  filterOverrides' =
+  filterOverrides' = filterOverrides'' null;
+  filterOverrides'' = excludePrio:
     let
       getPrio =
         def: if def.value._type or "" == "override" then def.value.priority else defaultOverridePriority;
@@ -1423,14 +1466,14 @@ let
     in
     defs:
     # Optimize for the singleton case, equivalent to the `else` clause.
-    if length defs == 1 then
+    if excludePrio == null && length defs == 1 then
       {
         values = map strip defs;
         highestPrio = getPrio (head defs);
       }
     else
       let
-        highestPrio = foldl' (prio: def: min (getPrio def) prio) 9999 defs;
+        highestPrio = foldl' (prio: def: min (getPrio def) prio) 9999 (if excludePrio == null then defs else filter (def: getPrio def != excludePrio) defs);
       in
       {
         values = concatMap (def: if getPrio def == highestPrio then [ (strip def) ] else [ ]) defs;
@@ -1462,7 +1505,11 @@ let
         else
           def;
       defs' = map strip defs;
-      compare = a: b: (a.priority or defaultOrderPriority) < (b.priority or defaultOrderPriority);
+      # mkApply mappers are sorted after other elements of the same priority.
+      compare = a: b: let # lessThan
+        a' = a.priority or defaultOrderPriority;
+        b' = b.priority or defaultOrderPriority;
+      in a' < b' || (a' == b' && (a.value._type or "" != "apply" && b.value._type or "" == "apply"));
     in
     sort compare defs';
 
@@ -1606,6 +1653,11 @@ let
   defaultOrderPriority = 1000;
   mkAfter = mkOrder 1500;
 
+  mkApply = mapper:
+    { _type = "apply";
+      inherit mapper;
+    };
+
   # Convenient property used to transfer all definitions and their
   # properties from one option to another. This property is useful for
   # renaming options, and also for including properties from another module
@@ -2317,6 +2369,7 @@ private
     mkAliasIfDef
     mkAliasOptionModule
     mkAliasOptionModuleMD
+    mkApply
     mkAssert
     mkBefore
     mkChangedOptionModule
diff --git a/lib/tests/modules.sh b/lib/tests/modules.sh
index 94aeea299e04..4a2ca2c171cb 100755
--- a/lib/tests/modules.sh
+++ b/lib/tests/modules.sh
@@ -168,6 +168,8 @@ checkConfigError() {
     fi
 }
 
+checkConfigOutput '^true$' config.result ./mkApply.nix
+
 # Shorthand meta attribute does not duplicate the config
 checkConfigOutput '^"one two"$' config.result ./shorthand-meta.nix
 
diff --git a/lib/tests/modules/mkApply.nix b/lib/tests/modules/mkApply.nix
new file mode 100644
index 000000000000..89ab617e8344
--- /dev/null
+++ b/lib/tests/modules/mkApply.nix
@@ -0,0 +1,156 @@
+#!/usr/bin/env -S bash -c 'bash <( tail -n+3 "$0" ) "$0" "$@"'
+/* begin script:
+
+nix-instantiate --timeout 1 --eval-only --show-trace --read-write-mode --json -E "$script" -A "$attr"
+
+exit # end script, begin document: */
+
+{ lib, config, ... }: let
+
+  appendToAll = suffix: (list: lib.map (x: x + suffix) list);
+
+  tests = {
+    array = { # listOf uses the v2 combined merge and check
+      type = lib.types.listOf lib.types.str;
+      apply = builtins.sort builtins.lessThan;
+      expected = [ "a!" "b!" "c!" "d!" ];
+      definitions = [
+        ["a"]
+        (lib.mkApply (appendToAll "!"))
+        ["b" "d"]
+        ["c"]
+      ];
+    };
+    ordered = {
+      type = lib.types.listOf lib.types.str;
+      apply = builtins.sort builtins.lessThan;
+      expected = [ "a!+?" "b!+?" "c+?" "d?" ];
+      definitions = [
+                      ["c"]
+        (lib.mkAfter  (lib.mkApply (appendToAll "?")))
+                      (lib.mkApply (appendToAll "+"))
+        (lib.mkBefore (lib.mkApply (appendToAll "!")))
+        (lib.mkBefore ["a" "b"])
+        (lib.mkAfter  ["d"])
+      ];
+    };
+    no-merge-v2 = { # as of writing, does not use v2 merging
+      type = lib.types.commas;
+      expected = "([{b},c],d)";
+      definitions = [
+                      "c"
+        (lib.mkAfter  (lib.mkApply (s: "(${s})")))
+                      (lib.mkApply (s: "[${s}]"))
+        (lib.mkBefore (lib.mkApply (s: "{${s}}")))
+        (lib.mkBefore "b")
+        (lib.mkAfter  "d")
+      ];
+    };
+    delete-attr = {
+      type = lib.types.attrsOf lib.types.anything;
+      expected = { good-entry = true; };
+      definitions = [
+        { good-entry = true; must-be-gone = true; }
+        { must-be-gone = lib.mkForce null; } # can't delete entry
+        (lib.mkApply (a: builtins.removeAttrs a [ "must-be-gone" ]))
+      ];
+    };
+    change-default = {
+      type = lib.types.str;
+      expected = "good";
+      definitions = [ # Change a default without affecting its priority:
+        (lib.mkOptionDefault "bad")
+        (lib.mkOptionDefault (lib.mkApply (value: "good")))
+        #(lib.mkOptionDefault "good") # Would not work: "The option `...' has conflicting definition values"
+        #(lib.mkDefault "good") # Would not work: changes priority
+      ];
+    };
+    change-default-override = {
+      type = lib.types.str;
+      expected = "custom";
+      definitions = [ # A changed default can be overridden:
+        (lib.mkOptionDefault "bad")
+        (lib.mkOptionDefault (lib.mkApply (value: "good")))
+        (lib.mkDefault "custom") # no conflict
+      ];
+    };
+    fallback-default = {
+      type = lib.types.listOf lib.types.str;
+      apply = builtins.sort builtins.lessThan;
+      expected = [ "good!" "more!" "yes!" ];
+      definitions = [ # When no value is set on the level that mkApply is used on, use the next lower priority level as input:
+        (lib.mkOptionDefault ["no"])
+        (lib.mkApply (appendToAll "!"))
+        (lib.mkIf true (lib.mkDefault ["yes"]))
+        (lib.mkMerge [ (lib.mkDefault ["more"]) (lib.mkDefault ["good"]) ])
+        (lib.mkIf false (lib.mkDefault ["never"]))
+      ];
+    };
+    fallback-default-override = {
+      type = lib.types.listOf lib.types.str;
+      apply = builtins.sort builtins.lessThan;
+      expected = [ "more!" "yes!" ];
+      definitions = [ # When no value is set on the level that mkApply is used on, use the next lower priority level as input:
+        (lib.mkOptionDefault ["no"])
+        (["yes"])
+        (lib.mkForce (lib.mkApply (appendToAll "!")))
+        (["more"])
+      ];
+    };
+    extend-script = {
+      type = lib.types.path;
+      expected = "/nix/store/wrapped";
+      definitions = [ #
+        "/nix/store/default"
+        (lib.mkApply (prev: "/nix/store/wrapped"))
+        #(lib.mkApply (prev: pkgs.writeShellScript "wrapped" ''
+        #  echo preparing ; source ${prev} ; echo done
+        #''))
+      ];
+    };
+    remove-package = {
+      type = lib.types.listOf (lib.types.attrsOf lib.types.anything);
+      apply = builtins.sort (a: b: throw "can't sort");
+      # programs.foo.enable = true;
+      # users.bar.packages = [ config.programs.foo.package ];
+      # environment.systemPackages = lib.mkApply (lib.remove config.programs.foo.package);
+      expected = [ { pname = "baz"; } ];
+      definitions = let
+        bar = { pname = "bar"; };
+        baz = { pname = "baz"; };
+      in [
+        [bar] [baz]
+        (lib.mkApply (lib.remove bar))
+      ];
+    };
+  };
+
+  importTest = name: args@{
+    type, definitions,
+    apply ? null,
+    ok ? (value: value == expected),
+    expected ? throw "missing default",
+  }: {
+    _file = "case '${name}' option";
+    options.${name} = lib.mkOption ({ inherit type; } // lib.optionalAttrs (args?apply) { inherit apply; });
+    options."${name}.ok" = lib.mkOption {
+      type = lib.types.bool; readOnly = true;
+      default = if ok config.${name} then true else lib.warn ''
+        Expected "${name}"
+        to be   ${lib.generators.toPretty { } expected},
+        but got ${lib.generators.toPretty { } config.${name}}
+      '' false;
+    };
+    imports = lib.imap1 (i: value: {
+      _file = "case '${name}' definition ${toString i}";
+      config.${name} = value;
+    }) definitions;
+  };
+
+in {
+  imports = builtins.attrValues (builtins.mapAttrs importTest tests);
+  options.result = lib.mkOption {
+    type = lib.types.bool; readOnly = true;
+    default = builtins.all (name: config."${name}.ok") (builtins.attrNames tests);
+  };
+}
